protocols/identify/src/behaviour.rs

```diff
@@ -274,7 +273,20 @@ impl NetworkBehaviour for Behaviour {
 
                 let observed = info.observed_addr.clone();
                 self.events
-                    .push_back(ToSwarm::GenerateEvent(Event::Received { peer_id, info }));
+                    .push_back(ToSwarm::GenerateEvent(Event::Received {
+                        peer_id,
+                        info: info.clone(),
+                    }));
+
+                for address in filter_new_external_addrs(
+                    &self.discovered_peers.get(&peer_id),
+                    &info.listen_addrs,
+                ) {
+                    self.events.push_back(ToSwarm::NewExternalAddrOfPeer {
+                        peer_id,
+                        addr: address.clone(),
+                    });
+                }
```

Comment:
So instead of filtering here I'd suggest the following:

- As commented on `PeerAddresses`, change the API of `put` to only accept a single address.
- Call `self.discovered_peers.add` for each address
- If that returns `true`, append an event to `self.events`

This avoids the filtering altogether.

Reply:
:heavy_check_mark: However maybe it would be good to merge this logic with what's happening in lines 271-272? Just to use `add` (instead of `put`) and check what was returned from `PeerAddresses`